content/en/blog/_posts/2019-03-27-kube-proxy-subtleties-debugging-an-intermittent-connection-resets.md

```diff
@@ -0,0 +1,169 @@
+
+Enough of the background, so what really went wrong and caused unexpected
+connection reset?
+
+As the diagram below shows, the problem is packet 3. Sometimes Conntrack cannot
+recognize a returning packet, and mark them as *INVALID*. The most common
+reasons include: Conntrack cannot keep track of a connection because it is out
+of capacity, the packet itself is out of a TCP window, etc. For those packets
+that have been marked as *INVALID* state by conntrack, we don’t have the
+iptables rule to drop it, so it will be forwarded to client pod, with source IP
+address not rewritten (as shown in packet 4)! Client pod doesn’t recognize this
+packet because it has a different source IP, which is pod IP, not service IP. As
+a result, client pod says, "Wait a second, I don't recall this connection to
+this IP ever exists, why this dude keeps sending this packet to me?" Basically
+what the client does, is simply sending a RST packet to the server pod IP, which
+is packet 5. Unfortunately, this is a totally legit pod-to-pod packet, which can
+be delivered to server pod. Server pod doesn’t know all the address translation
+that happens on the client side. From its view, packet 5 is totally legit
+packet, like packet 2 and 3. All server pod knows is, "Well, client pod doesn’t
+want to talk to me, so let’s close the connection!" Boom! Of course, in order
+for all these to happen, the RST packet has to be legit too, with the right TCP
+sequence number, etc. But when it happens, both parties agree to close the
+connection.
+
+{{<figure width="600"
+src="/images/blog/2019-03-26-kube-proxy-subtleties-debugging-an-intermittent-connection-resets/connection-reset-packet-flow.png"
+caption="Connection reset packet flow">}}
+
+## How to address it?
+
+Once we understand the root cause, the fix is not hard. There are at least 2
+ways to address it.
+
+- Make conntrack more liberal on packets, and don’t mark the packets as
+  *INVALID*. In Linux, you can do this by `echo 1 >
+  /proc/sys/net/ipv4/netfilter/ip_conntrack_tcp_be_liberal`.
+- Specifically add an iptables rule to drop the packets that are marked as
+  *INVALID*, so it won’t reach to client pod and cause harm.
+
+The fix is drafted (https://github.com/kubernetes/kubernetes/pull/74840), but
+unfortunately it didn’t catch the v1.14 release window. However, for the users
+that are affected by this bug, there is a way to mitigate the problem by apply
+the following rule in your cluster.
+
+```yaml apiVersion: extensions/v1beta1 kind: DaemonSet metadata: name:
+startup-script labels: app: startup-script spec: template: metadata: labels:
+app: startup-script spec: hostPID: true containers:
+      - name: startup-script image: gcr.io/google-containers/startup-script:v1
+        imagePullPolicy: IfNotPresent securityContext: privileged: true env:
+        - name: STARTUP_SCRIPT value: | #! /bin/bash echo 1 >
+          /proc/sys/net/ipv4/netfilter/ip_conntrack_tcp_be_liberal echo done ```
+
```

Comment:
could you please format the manifest file

Reply:
Sorry. My vim `gqG` totally screwed this up.